### PR TITLE
[llava][5/N] Add Llava model definition

### PR DESCRIPTION
**Pull request by larryliu0820**
Stack from [ghstack](https://github.com/ezyang/ghstack) (oldest at bottom):
* #4295
* __->__ #4259
* #4256

Summary:

Test Plan:

Reviewers:

Subscribers:

Tasks:

Tags:

Differential Revision: [D59759978](https://our.internmc.facebook.com/intern/diff/D59759978)

### CODE REVIEW

**Comment by larryliu0820** [comment]
@larryliu0820 has imported this pull request.  If you are a Meta employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D59759978).

**Comment by larryliu0820** [comment]
@larryliu0820 has imported this pull request.  If you are a Meta employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D59759978).

**Comment by larryliu0820** [comment]
@larryliu0820 has imported this pull request.  If you are a Meta employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D59759978).

**Comment by larryliu0820** [comment]
@larryliu0820 has imported this pull request.  If you are a Meta employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D59759978).

**Comment by larryliu0820** [comment]
@larryliu0820 has imported this pull request.  If you are a Meta employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D59759978).

**Comment by larryliu0820** [comment]
@larryliu0820 has imported this pull request.  If you are a Meta employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D59759978).

**Comment by larryliu0820** [comment]
@larryliu0820 has imported this pull request.  If you are a Meta employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D59759978).

**Comment by larryliu0820** [comment]
@larryliu0820 has imported this pull request.  If you are a Meta employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D59759978).

**Comment by larryliu0820** [comment]
@larryliu0820 has imported this pull request.  If you are a Meta employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D59759978).

**Comment by larryliu0820** [comment]
@larryliu0820 has imported this pull request.  If you are a Meta employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D59759978).

**Comment by larryliu0820** [comment]
@larryliu0820 has imported this pull request.  If you are a Meta employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D59759978).

**Comment by larryliu0820** [comment]
@larryliu0820 has imported this pull request.  If you are a Meta employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D59759978).

**Comment by larryliu0820** [comment]
@larryliu0820 has imported this pull request.  If you are a Meta employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D59759978).

File: examples/models/llava/install_requirements.sh [hunk 1]
```diff
@@ -17,8 +17,14 @@ pip install protobuf
 # Reinstall bitsandbytes to make it compatible.
 pip install bitsandbytes -I
 
+# numpy needs to be pin to 1.24. 1.26.4 will error out
```
Comment (helunwencser) [inline examples/models/llava/install_requirements.sh:20]: missing period.

File: examples/models/llava/install_requirements.sh [hunk 2]
```diff
@@ -17,8 +17,14 @@ pip install protobuf
...
+pip install numpy==1.24
+
+# Newer transformer will give TypeError: LlavaLlamaForCausalLM.forward() got an unexpected keyword argument 'cache_position'
```
Comment (helunwencser) [inline examples/models/llava/install_requirements.sh:23]: missing period.

File: examples/models/llava/model.py [hunk 3]
```diff
@@ -4,49 +4,373 @@
 # This source code is licensed under the BSD-style license found in the
 # LICENSE file in the root directory of this source tree.
 
+# An ExecuTorch friendly implementation of Llava-1.5.abs
```
Comment (helunwencser) [inline examples/models/llava/model.py:7]: missing period.

File: examples/models/llava/main.py [hunk 4]
```diff
@@ -0,0 +1,56 @@
...
+from model import LlavaModel
+
+
+def main():
+
+    llava_model = LlavaModel()
+    llava = llava_model.get_eager_model()
+
+    prompt_before_image, resized, prompt_after_image = llava_model.get_example_inputs()
+    print(f"Prompt: {llava_model.prompt}")
```
Comment (helunwencser) [inline examples/models/llava/main.py:18]: use logging instead?

File: examples/models/llava/main.py [hunk 5]
```diff
@@ -0,0 +1,56 @@
...
+            do_sample=False,
+            num_beams=1,
+            max_new_tokens=10,
+            use_cache=True,
+        )
+
+    outputs = llava_model.tokenizer.batch_decode(output_ids, skip_special_tokens=True)[
+        0
+    ].strip()
+    print(f"Reference output: {outputs}")
```
Comment (helunwencser) [inline examples/models/llava/main.py:34]: same

File: examples/models/llava/model.py [hunk 6]
```diff
@@ -4,49 +4,373 @@
...
+            self.prompt_after_image,
+        )
+        return self.input
+
+    def get_dynamic_shapes(self):
+        return self._get_image_dynamic_shapes()
+
+    def _get_image_dynamic_shapes(self):
+        height = Dim("height", min=8, max=336)
+        # height = Dim("height", min=8, max=4091)
```
Comment (helunwencser) [inline examples/models/llava/model.py:365]: why keep these commented code here?

**Review by helunwencser — APPROVED** [review]
LGTM. Had some minor comments.

**Comment by larryliu0820** [comment]
@larryliu0820 has imported this pull request.  If you are a Meta employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D59759978).

**Comment by larryliu0820** [comment]
@larryliu0820 has imported this pull request.  If you are a Meta employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D59759978).

**Comment by larryliu0820** [comment]
@larryliu0820 has imported this pull request.  If you are a Meta employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D59759978).

**Comment by larryliu0820** [comment]
@larryliu0820 has imported this pull request.  If you are a Meta employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D59759978).

**Pull request closed without merging**